### ISSUE TITLE
[Bug]: VLLM_BATCH_INVARIANT does not cover convolution, used in VAE blocks

### ISSUE THREAD

**Issue by ZJLi2013** [vllm-project/vllm#52155]
### Your current environment

OS                           : Ubuntu 22.04.5 LTS (x86_64)
Clang version                : 22.0.0git (roc-7.2.3 26084 f58b06dce1f9c15707c5f808fd002e18c2accf7e)
Libc version                 : glibc-2.35

PyTorch version              : 2.11.0+gitd0c8b1f
ROCM used to build PyTorch   : 7.2.53211
Python version               : 3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0] (64-bit runtime)
Python platform              : Linux-5.15.0-131-generic-x86_64-with-glibc2.35

Is CUDA available            : True
GPU models and configuration :  (gfx942:sramecc+:xnack-)   # 8x AMD Instinct MI300X
HIP runtime version          : 7.2.53211
MIOpen runtime version       : 3.5.1

CPU                          : Intel(R) Xeon(R) Platinum 8480C, 224 CPU(s), 2 sockets, 2 NUMA nodes

### 🐛 Describe the bug

`init_batch_invariance()` does not address convolution, so a model whose forward
contains one keeps a nondeterministic reduction while `VLLM_BATCH_INVARIANT=1` is on.

The `torch.library` override set covers `mm`, `addmm`, `matmul`, `linear`, `softmax`,
`mean.dim`, `bmm` and `rms_norm`. The only line touching conv is

```python
# vllm/model_executor/layers/batch_invariant.py, init_batch_invariance()
torch.backends.cudnn.conv.fp32_precision = "ieee"
```

which constrains precision, not reduction order — and `torch.backends.cudnn.deterministic`
is never set. Affected models are the ones with a convolution in the forward path:
multimodal vision patch-embeds (Qwen2.5-VL's is a `conv3d`) and diffusion VAEs. **This is
not ROCm-specific**: that flag exists because cuDNN also has nondeterministic convolution
algorithms, and it is unset on both backends. Possibly what
[#27059](https://github.com/vllm-project/vllm/issues/27059) hit before it was closed as
stale, though I have not verified that on CUDA.

**Suggested fix.** Set it alongside the existing `fp32_precision` lines — already inside
`if envs.VLLM_BATCH_INVARIANT:`, so the default path is untouched, and it has no effect
on models without convolutions. On ROCm the same flag maps to MIOpen, no backend branch.

**Scope.** The missing conv coverage is a code fact and platform-independent. The
measurement is run-to-run at fixed shape on ROCm, not batch invariance, and not a claim
about SM80/SM90. `cudnn.deterministic` also does not pin *which* algorithm is selected,
and the heuristics read batch size as part of the shape — deterministic conv and
batch-invariant conv stay different claims.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by sylvesterkaczmarek** [comment]
I’ve reviewed this against current `main` and would like to take it.

The minimal fix is to enable deterministic convolution algorithm selection inside the existing `VLLM_BATCH_INVARIANT` initialization, alongside the current precision controls:

```python
if envs.VLLM_BATCH_INVARIANT:
    override_envs_for_invariance()
    enable_batch_invariant_mode()
    torch.backends.cudnn.deterministic = True

    torch.backends.cuda.matmul.fp32_precision = "ieee"
    torch.backends.cudnn.conv.fp32_precision = "ieee"
    torch.backends.cudnn.rnn.fp32_precision = "ieee"
```

I also have a focused regression test prepared that forces `VLLM_BATCH_INVARIANT=True`, invokes `init_batch_invariance()`, verifies `torch.backends.cudnn.deterministic` becomes true, and restores the global backend state afterward.

This keeps the default path unchanged and addresses deterministic convolution algorithm selection without claiming that it alone proves full batch-shape invariance for every convolution implementation.

**Comment by sylvesterkaczmarek** [comment]
I’ve reviewed this against current `main` and would like to take it.

The minimal fix is to enable deterministic convolution algorithm selection inside the existing `VLLM_BATCH_INVARIANT` initialization, alongside the current precision controls:

```python
if envs.VLLM_BATCH_INVARIANT:
    override_envs_for_invariance()
    enable_batch_invariant_mode()
    torch.backends.cudnn.deterministic = True

    torch.backends.cuda.matmul.fp32_precision = "ieee"
    torch.backends.cudnn.conv.fp32_precision = "ieee"
    torch.backends.cudnn.rnn.fp32_precision = "ieee"
```

I also have a focused regression test prepared that forces `VLLM_BATCH_INVARIANT=True`, invokes `init_batch_invariance()`, verifies `torch.backends.cudnn.deterministic` becomes true, and restores the global backend state afterward.

This keeps the default path unchanged and addresses deterministic convolution algorithm selection without claiming that it alone proves full batch-shape invariance for every convolution implementation.
